sycl/doc/extensions/SPIRV/SPV_INTEL_arbitrary_precision_floating_point.asciidoc

```diff
@@ -0,0 +1,1280 @@
+= SPV_INTEL_arbitrary_precision_floating_point
+
+== Name Strings
+
+SPV_INTEL_arbitrary_precision_floating_point
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/KhronosGroup/SPIRV-Headers
+
+== Contributors
+
+Ajaykumar Kannan, Intel
+
+== Notice
+
+Copyright (c) 2020 Intel Corporation.  All rights reserved.
+
+== Status
+
+Initial Release
+
+== Version
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | {docdate}
+| Revision           | 1
+|========================================
+
+== Dependencies
+
+This extension is written against the SPIR-V Specification Version 1.5, Revision 2.
+
+This extension depends on the SPV_INTEL_arbitrary_precision_int spec.
+
+== Overview
+
+_ihc_float_ is a datatype represented in SPIR-V using `OpTypeInt` that implements arbitrary precision floating point support.
```

Comment:
It seems a bit strange that arbitrary precision floating-point type is built on top of integer type. I guess  [`OpTypeFloat`](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpTypeFloat) can't be really used here, as it requires IEEE754 compliance - right?

> The bit pattern of a floating-point value is as described by the IEEE 754 standard.

I don't have strong objections against it, since the only thing you need is just a bits storage - all the semantic is encoded in operations and not in the data types, but still, this looks strange.

I think that having separate type would be better here: there is a term ["Floating-point type"](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Floating) in the spec, and these new arbitrary precision floating-point types **are not** floating-point types at all from the spec point of view. If you had them declared as floating-point types with some requirements to bit pattern (by referring to some other standard/spec or describing it here directly), you could easily get support for other arithmetic operations like [`OpFMul`](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpFMul), for example. Note that its description doesn't refer to `OpTypeFloat`, but rather refers to `Floating-point type`, which means that it could be used for other floating-point types without adding new instructions or updating any wording.

Reply:
> I'm not sure what the Width of the OpTypeFloat would imply about the size of the mantissa and exponents. Perhaps it was originally designed for the standard floating point types (e.g. half, single, double precision floats)?

Yes, it was for sure designed for standard floating-point types

> To your note, these operations do follow a type that is aligned very well with the IEEE spec, just that they provide the capability of using any arbitrary-sized mantissa/exponent, but the representation is in line with IEEE. So perhaps we should switch to using the OpTypeFloat with an extra parameter to clarify the size of the mantissa/exponent?

I think it should be okay to just use `OpTypeFloat` so new functions operate on actual floating-point type. Regarding encoding the size of the mantissa/exponent: it is probably okay to encode this in the operation itself. I.e. we re-use existing data type, which is in line with IEEE, but we provide new instructions that re-interpret it in a bit different way (length of mantissa/exponent is not a standard one)

Reply:
In IR these parameters types are represented via ExtInt keyword. It may be confusing for a user that arbitrary length int is translated into SPIR-V as float type.

Reply:
I'll leave this as is for now. @AlexeySachkov what do you think?

Reply:
> It may be confusing for a user that arbitrary length int is translated into SPIR-V as float type.

I'm a bit confused: this file states that arbitrary precision float is translated into SPIR-V as int type.

My suggestion were to align that, but I think it is not critical to leave that as-is